ember-client/shared/src/main/scala/org/http4s/ember/client/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,110 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.ember.client.internal
+
+import cats.MonadThrow
+import cats.data.NonEmptyList
+import cats.syntax.all._
+import org.http4s.Status
+import org.http4s._
+import org.http4s.crypto.Hash
+import org.http4s.crypto.HashAlgorithm
+import org.http4s.headers.Connection
+import org.http4s.headers._
+import org.http4s.websocket.Rfc6455
+import org.typelevel.ci._
+import scodec.bits.ByteVector
+
+private[client] object WebSocketHelpers {
+
+  private[internal] val supportedWebSocketVersion = 13L
+
+  private[internal] val upgradeCi = ci"upgrade"
+  private[internal] val webSocketProtocol = Protocol(ci"websocket", None)
+  private[internal] val connectionUpgrade = Connection(NonEmptyList.of(upgradeCi))
+  private[internal] val upgradeWebSocket = Upgrade(webSocketProtocol)
+
+  /** Validate the opening handshake response from the server
+    * https://datatracker.ietf.org/doc/html/rfc6455#page-6
+    */
+  def validateServerHandshake[F[_]](
+      response: Response[F],
+      secWebSocketKey: String,
+  )(implicit F: MonadThrow[F]): F[Either[ServerHandshakeError, Unit]] =
+    for {
+      secWebSocketAccept <- serverHandshake(response).pure[F]
+      correctSecWebSocketAccept <- clientHandshake(secWebSocketKey)
+      validated = secWebSocketAccept.flatMap(s =>
+        if (s == correctSecWebSocketAccept) Either.unit else Left(InvalidSecWebSocketAccept)
+      )
+    } yield validated
+
+  private[this] val magic = ByteVector.view(Rfc6455.handshakeMagicBytes)
+
+  private[internal] def clientHandshake[F[_]](
+      value: String
+  )(implicit F: MonadThrow[F]): F[ByteVector] = for {
+    value <- ByteVector.encodeAscii(value).liftTo[F]
+    digest <- Hash[F].digest(HashAlgorithm.SHA1, value ++ magic)
+  } yield digest
+
+  private def serverHandshake[F[_]](res: Response[F]): Either[ServerHandshakeError, ByteVector] = {
+    val status = res.status match {
+      case Status.SwitchingProtocols => Either.unit
+      case _ => Left(InvalidStatus)
+    }
+
+    val connection = res.headers.get[Connection] match {
+      case Some(header) if header.hasUpgrade => Either.unit
+      case _ => Left(UpgradeRequired)
+    }
+
+    val upgrade = res.headers.get[Upgrade] match {
+      case Some(header) if header.values.contains_(webSocketProtocol) => Either.unit
+      case _ => Left(UpgradeRequired)
+    }
+
+    val secWebSocketAcceptKey = res.headers.get[`Sec-WebSocket-Accept`] match {
+      case Some(header) => Right(header.hashedKey)
+      case None => Left(SecWebSocketAcceptNotFound)
+    }
+
+    (status, connection, upgrade, secWebSocketAcceptKey).mapN {
+      case (_, _, _, secWebSocketAcceptKey) =>
+        secWebSocketAcceptKey
+    }
```

Comment:
This way each check evaluates lazily and will short-circuit immediately if there is an error.

```suggestion
  private def serverHandshake[F[_]](res: Response[F]): Either[ServerHandshakeError, ByteVector] = {
    def status = res.status match {
      case Status.SwitchingProtocols => Either.unit
      case _ => Left(InvalidStatus)
    }

    def connection = res.headers.get[Connection] match {
      case Some(header) if header.hasUpgrade => Either.unit
      case _ => Left(UpgradeRequired)
    }

    def upgrade = res.headers.get[Upgrade] match {
      case Some(header) if header.values.contains_(webSocketProtocol) => Either.unit
      case _ => Left(UpgradeRequired)
    }

    def secWebSocketAcceptKey = res.headers.get[`Sec-WebSocket-Accept`] match {
      case Some(header) => Right(header.hashedKey)
      case None => Left(SecWebSocketAcceptNotFound)
    }

    status >> connection >> upgrade >> secWebSocketAcceptKey
```